src/dc_tools.rs

```diff
@@ -1528,29 +1538,57 @@ impl<T: AsRef<std::ffi::OsStr>> OsStrExt for T {
 #[allow(dead_code)]
 fn os_str_to_c_string_unicode(
     os_str: &dyn AsRef<std::ffi::OsStr>,
-) -> Result<std::ffi::CString, CStringError> {
+) -> Result<CString, CStringError> {
     match os_str.as_ref().to_str() {
-        Some(val) => std::ffi::CString::new(val.as_bytes()).map_err(|err| match err {
+        Some(val) => CString::new(val.as_bytes()).map_err(|err| match err {
             std::ffi::NulError { .. } => CStringError::InteriorNullByte,
         }),
         None => Err(CStringError::NotUnicode),
     }
 }
 
-pub fn to_cstring<S: AsRef<str>>(s: S) -> std::ffi::CString {
-    std::ffi::CString::new(s.as_ref()).unwrap()
+/// Needs to free the result after use!
+pub unsafe fn to_cstring<S: AsRef<str>>(s: S) -> *mut libc::c_char {
+    let cstr = CString::new(s.as_ref()).expect("invalid string converted");
+    dc_strdup(cstr.as_ref().as_ptr())
 }
 
 pub fn to_string(s: *const libc::c_char) -> String {
     if s.is_null() {
         return "".into();
     }
-    unsafe { std::ffi::CStr::from_ptr(s).to_str().unwrap().to_string() }
+
+    let cstr = unsafe { CStr::from_ptr(s) };
```

Comment:
Should the entire function not be unsafe?  Likewise for to_string_lossy() and as_str().  Although It's probably best to make such a change in a separate PR.